### PR TITLE
feat: Add Cohere reranker config

### PR DESCRIPTION
**Pull request by leehuwuj**
Resolve: https://github.com/ragapp/ragapp/issues/107

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -3,21 +3,25 @@
 
 load_dotenv(dotenv_path=ENV_FILE_PATH, verbose=False)
 
-import os
+# flake8: noqa
```
Comment (marcusschiesser) [inline main.py:6]: why we need this?

File: patch/backend/app/engine/reranker.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import os
+
+from llama_index.core.postprocessor.types import BaseNodePostprocessor
+
+
+def get_cohere_reranker():
+    from llama_index.postprocessor.cohere_rerank import CohereRerank
+
+    api_key = os.getenv("COHERE_API_KEY")
+    top_k = int(os.getenv("TOP_K", "3"))
```
Comment (marcusschiesser) [inline patch/backend/app/engine/reranker.py:10]: redundant with engine - centralize?

Comment (leehuwuj) [reply patch/backend/app/engine/reranker.py:10]: Let me check to see is it able to centralize, but the original idea was the engine use it for both retrieval while this is for reranker, we just share the TOP_K as the expected number of result - no matter the reranker is enable or not.

Comment (marcusschiesser) [reply patch/backend/app/engine/reranker.py:10]: good point, but then we can do this?
```suggestion
    top_k = int(os.getenv("TOP_K", DEFAULT_TOP_K))
```

File: patch/backend/app/engine/__init__.py [hunk 3]
```diff
@@ -1,13 +1,24 @@
...
 
 
 def get_chat_engine():
-    top_k = int(os.getenv("TOP_K", "3"))
     system_prompt = os.getenv("SYSTEM_PROMPT")
+    node_postprocessors = []
+
+    top_k = int(os.getenv("TOP_K", DEFAULT_TOP_K))
+    if os.getenv("RERANK_PROVIDER") is not None:
+        top_k = max(top_k, DEFAULT_MAX_TOP_K)
```
Comment (marcusschiesser) [inline patch/backend/app/engine/__init__.py:20]: own env variable? maybe called `RERANK_TOP_K`?

File: admin/sections/config/reranker.tsx [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                      field.onChange(checked);
+                      submitRerankerForm(form.getValues());
+                    }}
+                  />
+                </FormControl>
+                <div>
+                  <FormLabel className="font-normal">Use Reranker</FormLabel>
+                  <FormMessage />
+                  <FormDescription>
+                    Use Cohere reranker to rerank the retrieved documents
```
Comment (marcusschiesser) [inline admin/sections/config/reranker.tsx:81]: I guess we might add more rerankers later, how about:
```suggestion
                    Use a reranker to improve the accuracy of the retrieved documents
```

File: admin/sections/config/reranker.tsx [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                    <a
+                      href="https://dashboard.cohere.com/api-keys"
+                      target="_blank"
+                      rel="noreferrer"
+                    >
+                      https://dashboard.cohere.com/api-keys
+                    </a>
+                  </FormDescription>
+                  <FormMessage />
+                </FormItem>
```
Comment (marcusschiesser) [inline admin/sections/config/reranker.tsx:110]: i think if a user is using a reranker, he's more serious, so he should be able to change TOP_K and RERANK_TOP_K

Comment (marcusschiesser) [reply admin/sections/config/reranker.tsx:110]: For now, i wouldn't generally allow changing TOP_K if the user doesn't select a reranker (complicates setup) - maybe we have an "Advanced" section later for that.

Comment (leehuwuj) [reply admin/sections/config/reranker.tsx]: That makes sense, i just added config to config RERANK_TOP_K

Comment (leehuwuj) [reply admin/sections/config/reranker.tsx]: > we just share the TOP_K as the expected number of result - no matter the reranker is enable or not.

After all, i think the point above could confuse the user if we add config for `TOP_K` later. So, i'd just use the TOP_K  and RERANK_TOP_K as the starter idea: If the reranker is enabled, the top_k would be changed to max(top_k, rerank_top_k) (larger value than the rerank_top_k). We should add a note for this if we add config for top_k later.

File: admin/client/reranker.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import { z } from "zod";
+import { getBaseURL } from "./utils";
+
+export const CohereRerankerConfigSchema = z.object({
+  use_reranker: z.boolean().optional(),
+  rerank_provider: z.string().trim().optional(),
+  cohere_api_key: z.string().trim().optional(),
+  top_k: z.coerce.number().int().optional().default(5),
```
Comment (marcusschiesser) [inline admin/client/reranker.ts]: needed?

**Review by marcusschiesser — APPROVED** [review]

**Pull request merged**